src/analysis-cache.ts

```diff
@@ -0,0 +1,169 @@
+/**
+ * @license
+ * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Document, Import, InlineDocument, ScannedDocument, ScannedImport, ScannedInlineDocument} from './model/model';
+import {ParsedDocument} from './parser/document';
+
+export class AnalysisCache {
+  parsedDocumentPromises = new Map<string, Promise<ParsedDocument<any, any>>>();
+  scannedDocumentPromises = new Map<string, Promise<ScannedDocument>>();
+  analyzedDocumentPromises = new Map<string, Promise<Document>>();
+  dependenciesScanned = new Map<string, Promise<void>>();
+
+  scannedDocuments = new Map<string, ScannedDocument>();
+  analyzedDocuments = new Map<string, Document>();
+
+  /**
+   * Returns a copy of this cache, with the given path and all of its transitive
+   * dependants invalidated.
+   */
+  onPathChanged(path: string, dependants: Iterable<string>): AnalysisCache {
+    const newCache = this._clone();
+    newCache.parsedDocumentPromises.delete(path);
+    newCache.scannedDocumentPromises.delete(path);
+    newCache.dependenciesScanned.delete(path);
+    newCache.scannedDocuments.delete(path);
+    newCache.analyzedDocuments.delete(path);
+
+    // Analyzed documents need to be treated more carefully, because they have
+    // relationships with other documents. So first we remove all documents
+    // which transitively import the changed document. We also need to mark
+    // all of those docs as needing to rescan their dependencies.
+    for (const invalidatedPath of dependants) {
+      newCache.dependenciesScanned.delete(invalidatedPath);
+      newCache.analyzedDocuments.delete(invalidatedPath);
+    }
+    // Then we clear out the analyzed document promises, which could have
+    // in-progress results that don't cohere with the state of the new cache.
+    // Only populate the new analyzed promise cache with results that are
+    // definite, and not invalidated.
+    newCache.analyzedDocumentPromises.clear();
+    for (const keyValue of newCache.analyzedDocuments) {
+      newCache.analyzedDocumentPromises.set(
+          keyValue[0], Promise.resolve(keyValue[1]));
+    }
+
+    return newCache;
+  }
+
+  private _clone(): AnalysisCache {
+    const newCache = new AnalysisCache();
```

Comment:
I'd use a copy constructor instead so you can use the copy constructor of `Map`:

```javascript
constructor(from?: AnalysisCache) {
  this.parsedDocumentPromises = new Map(from && from.parsedDocumentPromises);
  // etc...
}
```